### DISCUSSION TITLE
Updates on Query-Streamlink

### DISCUSSION THREAD

**Discussion by BellezaEmporium** [iptv-org/community#612] [Show and tell]
Edited on 26/03/2022

This is the discussion feed linked to [query-streamlink](https://github.com/LaneSh4d0w/query-streamlink), addressing the different updates for the services.

Right now, one of the most stable/best iterations of my program is @dct-infra 's amazing free.fullspeed.tv server.

I have seen multiple forks coming from the Latin Americas, having seen Heroku-backed services multiply themselves every day.
For this reason, I've rewrote the README in 3 different languages : English, Spanish and French.
Since my Spanish isn't the best, I've contacted a friend of mine who took the time to review my broken Spanish, thanks to her, the Spanish README should be good.

**WARNING**
I've reworked completely query-streamlink to be faster, less "tweaky" and more conventional, fixing things in the same instance. Streann links should now work flawlessly, as well as DASH streams (welcome back to certain streams that were locked due to issues related in this).

I've also forked the official Streamlink repo to create my own plugins, some that are actually used on @dct-infra 's free.fullspeed.tv server. Karwan, Giniko, OpenCaster (for local usage), ConceptoWeb and (soon) HTA / Shashat are such plugins that I've implemented.

A Django and Flask implementation of query-streamlink are available. Cherrypy implementation should be coming later.

**Comment by keystroke3** [comment]
I still haven't found a good proxy service yet. Implementing caching helped a lot with youtube, maybe you should try it in your instance if you get a new ip. It's always good to have multiple options in case one of our services is down.

**Reply by BellezaEmporium** [reply]
I'll have a look at that.

**Comment by BellezaEmporium** [comment]
Thanks to @keystroke3 's help on the code, we now have an API, for developers in need of links that they want to use on any kind of use-case. Note that this is still a work in progress, due to slight "quacks" that might happen on certain cases.

Code was refactored, and i'm trying to make it work on Vercel, so we'll be transitioning links.

**Comment by BellezaEmporium** [comment]
After some time, it's finally available ! https://query-streamlink.herokuapp.com

**Reply by keystroke3** [reply]
Ah, that's nice. At least there will be an alternative. I haven't had time to fix my implementation yet.

**Reply by BellezaEmporium** [reply]
Yeah. When I've found out the mistake was right in front of me (Procfile that was not correct), I was laughing at myself. Now everything works as expected. Limiter seems to be OK too.

**Comment by BellezaEmporium** [comment]
![image](https://user-images.githubusercontent.com/30985701/130479644-d8ecd96e-18bc-4195-8047-28ca5cd8bbba.png)

Interesting on how people could spam a single channel.

**Reply by keystroke3** [reply]
Yes, that's why I had redis to cache the requests.

**Reply by BellezaEmporium** [reply]
> 
> 
> Yes, that's why I had redis to cache the requests.

I'm kind of thinking I could add a kind of redis to my files. I think I could only apply it to YouTube, because the rest I don't know how that works.

**Reply by keystroke3** [reply]
If you don't feel like using redis, you could just use a python dictionary with the stream queried, the link for the stream and the expiry. 
```python
{
stream1:[link, expiry],
stream2:[link, expiry],
....
}
```
You can then store this data in a pickle. The program can then check to see if the the requested stream is in the 'cache' dictionary. If it is there, it checks to see if it has expired. If is still good, it returns the link, else it fetches a new link and updates the stream information. Every time a new stream query is made, the information is added in this dictionary. This solution is a bit slow as the pickle file has to be opened for every request that comes through, but is good when deploying on serverless platforms like Heroku. A redis server would still be way better, because it stores the data in memory and is much easier to query.

**Comment by BellezaEmporium** [comment]
- README's got 3 different languages now : English, French and Spanish (Spanish one is a WIP, my Spanish isn't quite good, luckily I got friends to address the different grammatical issues) to answer the new fork-o-rama ongoing from LATAM citizens.
- Rewrote index page to actually tell the end user how to use the program.
- I am planning to write the program for multiple Python webservices (per request from @dct-infra , to monitor the different webservices and see which one works better)

**Comment by BellezaEmporium** [comment]
Yesterday, I've initiated a major rework of query-streamlink. Following changes will be : 

- Fixed DASH streams, as they weren't working on base version (tested on French channel TF1).
- If a stream couldn't be catched, actually give a reason why it didn't work out.
- Removing "quality" parameter and API system as they weren't used.
- Simplified query-streamlink usage without breaking any new implementation.

This will affect READMEs too, as they'll need a bit of a tweak.

**Reply by BellezaEmporium** [reply]
New version is out, feel free to check.

**Comment by BellezaEmporium** [comment]
Recent updates : 

- Turkish README was added, including a rework of the main program (thanks to @[keyiflerolsun](https://github.com/keyiflerolsun))
- Various plugins were added : dacast, azro
- New plugins to come soon : teveo (cu/co), faulio (uses cilive)

**Comment by BellezaEmporium** [comment]
Been a while ! 

Since then, the code has been reoptimized, and the dependencies updated.
